---
title: "Lethality"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
params:
    results_dir: here("tests", "testthat", "results")
---


## Introduction

The purpose of this workflow is to look for potential 
genetic causes of a bimodal distribution in temperature at 
7 hours post LPS injection in MOLF/B6 heterozygotes. It is 
possible that there is a second locus that is driving this 
split.


```{r set_path, message = FALSE, error = FALSE, warning = FALSE}
rm(list = ls())

library(here)

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("pheatmap", "qtl2", "cluster", "DT", "glmnet")
load_libraries(all.packages)

load_latest_cape(here("../../../git_repositories/cape"))
```

```{r load_data}
lps <- read_cross2(here("Data", "qtl2_files_mod", "qtl2_files", "lps.yaml"))
lps_cape <- qtl2_to_cape(lps)
data_obj <- lps_cape$data_obj
geno_obj <- lps_cape$geno_obj

gene.info <- read.delim(here("Data", "general", "mouse_gene_info.txt"))
```

```{r plot_allele_effect}
plot_allele_effects <- function(plot_data_obj, marker.name, plot.ind = c("females", "males", "all"),
    plot_type = "p", ymin = NULL, ymax = NULL){
    
    has_covar <- as.logical(length(plot_data_obj$p_covar_table))
    if(!has_covar){
        sex.col <- plot_data_obj$pheno[,"sex"]
        plot_data_obj$pheno <- plot_data_obj$pheno[,1:8]
    }else{
        sex.col <- plot_data_obj$p_covar_table[,which(data_obj$p_covar == "sex")]
    }
    if(plot.ind == "females"){
        remove.ind <- which(sex.col == 1)
    }
    if(plot.ind == "males"){
        remove.ind <- which(sex.col == 0)
    }
    if(plot.ind == "all"){
        remove.ind = NULL
    }

    pared_obj <- suppressWarnings(remove_ind(plot_data_obj, remove.ind))

    plot_effects(pared_obj, geno_obj, marker1 = marker.name, marker2 = NULL, 
        plot_type = plot_type, error_bars = "se", pheno_type = "Normalized", 
        gen_model1 = "Additive", ymin = ymin, ymax = ymax)

}

nearest_marker <- function(data_obj, gene.name, gene.info){
    gene.idx <- which(gene.info[,"external_gene_name"] == gene.name)
    if(length(gene.idx) == 0){stop(paste("Cannot find", gene.name))}
    gene.chr <- gene.info[gene.idx, "chromosome_name"]
    gene.pos <- gene.info[gene.idx, "start_position"]
    chr.idx <- which(data_obj$chromosome == gene.chr)
    chr.pos <- data_obj$marker_location[chr.idx]
    nearest.idx <- get.nearest.pt(chr.pos, gene.pos)
    nearest.marker <- data_obj$geno_names[[3]][chr.idx[nearest.idx]]
    return(nearest.marker)
}

```

The following plot shows body temperature at each hour following LPS 
injection for females. There is a bimodal distribution in hour 7.

```{r female_split, fig.width = 10, fig.height = 6}

zbp1.marker <- nearest_marker(data_obj, "Zbp1", gene.info)
plot_allele_effects(data_obj, paste0(zbp1.marker, "_B"), 
    plot.ind = "females", plot_type = "p", ymin = 25, ymax = 45) #look at X7h
```

There is no such split in the males

```{r males, fig.width = 10, fig.height = 6}
plot_allele_effects(data_obj, paste0(zbp1.marker, "_B"), 
    plot.ind = "males", plot_type = "p", ymin = 25, ymax = 45) #look at X7h
```

We are interested in whether there is a genetic explanation for this.
That is, is there a second locus that is interacting with the Zbp1
locus to drive this split. 

We are going to check this as simply as possible with the following
procedure:

1) Isolate the individuals that have the bimodal split (female 
    mice heterozygous for Zbp1).
2) Treat the phenotype as high and low factors, and each marker
    genotype as a factor with two or three levels depending on the
    represented genotypes at that locus.
3) Perform a Fisher exact test on the contingency table of the 
    occurrence of each genotype with each phenotype. Markers
    that deviate from the expected distribution.

Markers that are correlated with this trait in these animals
may be contributing to the observed phenotype.

Below we pulled out the Zbp1 female heterozygotes. The plot
shows the split again. It looks as if 34 degrees is a good 
place to split the hets into high and low temperature animals.

```{r subset}
zbp1.geno <- round(geno_obj[,,which(data_obj$geno_names[[3]] == zbp1.marker)][,2]*100)/100

f.idx <- which(data_obj$pheno[,"sex"] == 0)
m.idx <- which(data_obj$pheno[,"sex"] == 1)
stripchart(data_obj$pheno[f.idx,"X7h"]~zbp1.geno[f.idx], vertical = TRUE, method = "jitter",
    pch = 16, col = "#ef8a62", group.names = c("B6 hom", "het", "MOLF hom"), 
    ylab = "Body Temperature")
abline(h = 34, lty = 2, col = "gray", lwd = 3)


het.idx <- which(zbp1.geno == 0.5)
high.temp.idx <- which(data_obj$pheno[f.idx,"X7h"] > 34)
low.temp.idx <- which(data_obj$pheno[f.idx,"X7h"] < 34)

high.het <- intersect(het.idx, f.idx[high.temp.idx])
low.het <- intersect(het.idx, f.idx[low.temp.idx])

f.het <- intersect(het.idx, f.idx)
het.pheno <- data_obj$pheno[f.het,"X7h",drop=FALSE]
```

The plot below verifies that we have pulled out the correct
individuals. There are `r length(f.het)` individuals in the
group overall. There are `r length(high.het)` animals in this
group with a high temperature, and `r length(low.het)` animals
in this group with a low temperature.

```{r verify}
boxplot(list("High Het" = data_obj$pheno[high.het,"X7h"], "Low Het" = data_obj$pheno[low.het,"X7h"]),
    ylab = "Body Temperature")
```

## Overall Results

The following plot shows the -log10(p values) for markers that 
have genotypes non-randomly distributed across the bimodal split. 
The horizontal line shows the significance at p = 0.05. There are 
quite a few peaks above the threshold.

```{r extract_geno, fig.width = 10, fig.height = 4, warning = FALSE, message = FALSE, error = FALSE}
#pull out the genotype cods for all markers
test.geno <- round(geno_obj[f.het,2,]*100)/100
#tests <- apply(test.geno, 2, function(x) lm(het.pheno~x))
#test.pval <- sapply(tests, lm_p)

#Fisher exact test
het.factor <- het.pheno
het.factor[which(het.pheno > 34)] <- "high"
het.factor[which(het.pheno < 34)] <- "low"
cont.tables <- apply(test.geno, 2, 
    function(x) table(data.frame("pheno" = as.factor(het.factor), "geno" = as.factor(x))))
tests <- lapply(cont.tables, function(x) if(ncol(x) > 1){fisher.test(x)}else{list("p.value" = NA)})
names(tests) <- colnames(test.geno)
test.pval <- sapply(tests, function(x) x$p.value)

#permutations
perm.file <- here("Results", "Zbp1_Fisher_perm.RDS")
if(!file.exists(perm.file)){
    nperm <- 100
    pval.perm <- matrix(NA, nrow = nperm, ncol = length(tests))
    colnames(pval.perm) <- names(tests)
    for(p in 1:nperm){
        #print(p)
        #randomize phenotype with respect to genoytpe
        perm.tables <- apply(test.geno, 2, 
            function(x) table(data.frame("pheno" = as.factor(het.factor), "geno" = as.factor(sample(x)))))
        perm.tests <- lapply(perm.tables, function(x) if(ncol(x) > 1){fisher.test(x)}else{list("p.value" = NA)})
        perm.p <- sapply(perm.tests, function(x) x$p.value)
        pval.perm[p,] <- perm.p
    }
    saveRDS(pval.perm, perm.file)
}else{
    pval.perm <- readRDS(perm.file)
}

nlog.thresh <- get.percentile(unlist(-log10(pval.perm)), 95)

plot_stats_on_chr(-log10(test.pval), data_obj$chromosome, 
    mark.y = nlog.thresh, y.label = "-log10(p value)")
```

## Effects of individual markers {.tabset .tabset-fade .tabset-pills}

The boxplots below show effects of markers on body temperature
only in the female mice heterozygous for B6/MOLF at the Zbp1 locus.
The break between the groups is marked by a horizontal line. The
results below are for markers that had a -log10(p value) greater
than `r nlog.thresh`.

There are many results, so only the top marker from each chromosome 
is shown.


```{r find_max, results = "asis"}
plot.all.animals = FALSE #Sasha asked for some specific plots for the Mov10 locus
                         #the code for that is wrapped in the plot.all.animals section

large.effect <- which(-log10(test.pval) > nlog.thresh)
marker.idx <- match(names(large.effect), data_obj$geno_names$locus)
large.effect.chr <- data_obj$chromosome[marker.idx]
u_chr <- unique(large.effect.chr)

#kleep track of maximum markers
max.chr <- vector(mode = "list", length = length(u_chr))
names(max.chr) <- u_chr
for(ch in 1:length(u_chr)){
    effect.chr <- u_chr[ch]
    max.effect <- which.max(-log10(test.pval[large.effect[which(large.effect.chr == effect.chr)]]))
    effect.idx <- which(data_obj$geno_names$locus == names(max.effect))
    names(effect.idx) <- names(max.effect)
    max.chr[[ch]] <- effect.idx
    max.loc <- data_obj$marker_location[effect.idx]
    cat("###", paste0("Chr", effect.chr, "; ", signif(max.loc/1e6, 3), "Mb\n"))
    boxplot(het.pheno~test.geno[,effect.idx], ylab = "Body Temperature",
        xlab = paste0("Genotype at Chr", effect.chr, "; ", signif(max.loc/1e6, 3), "Mb"))
    stripchart(het.pheno~test.geno[,effect.idx], vertical = TRUE, add = TRUE, col = "#ef8a62",
        pch = 16, method = "jitter")
    abline(h = 34)

    if(plot.all.animals){
        #all mice
        par(mfrow = c(1,2))
        all.pheno <- data_obj$pheno[,"X7h",drop=FALSE]
        all.geno <- as.factor(round(geno_obj[,2,effect.idx]*100)/100)
        model <- lm(all.pheno[f.idx]~all.geno[f.idx])
        pval <- anova(model)$"Pr(>F)"[1]
        plot(all.pheno[f.idx]~all.geno[f.idx],
            ylab = "Body Temperature", main = paste("Females\np =", signif(pval, 2)),
            xlab =  paste0("Chr", effect.chr, "; ", signif(max.loc/1e6, 3), "Mb\n"))
        stripchart(all.pheno[f.idx]~all.geno[f.idx],
            method = "jitter", pch = 16, col = "#ef8a62", vertical = TRUE, add = TRUE)
        abline(h = 34)

        #sub.test <- lapply(levels(all.geno[f.idx]), function(x) t.test(all.pheno[f.idx[which(all.geno[f.idx] == x)]], alternative = "greater", mu = 34))
        #sapply(sub.test, function(x) x$p.value)

        model <- lm(all.pheno[m.idx]~all.geno[m.idx])
        pval <- anova(model)$"Pr(>F)"[1]
        plot(all.pheno[m.idx]~all.geno[m.idx],
            ylab = "Body Temperature", main = paste("Males\np =", signif(pval, 2)),
            xlab =  paste0("Chr", effect.chr, "; ", signif(max.loc/1e6, 3), "Mb\n"))
        stripchart(all.pheno[m.idx]~all.geno[m.idx],
            method = "jitter", pch = 16, col = "#ef8a62", vertical = TRUE, add = TRUE)
        abline(h = 34)

        #sub.test <- lapply(levels(all.geno[m.idx]), function(x) t.test(all.pheno[m.idx[which(all.geno[f.idx] == x)]], alternative = "greater", mu = 34))
        #sapply(sub.test, function(x) x$p.value)


        #are high-temperature mice enriched in any of the genotypes?
        num.high <- sapply(c(0, 0.5, 1), function(x) length(intersect(which(all.geno[f.idx] == x), which(all.pheno[f.idx] > 34))))
        num.low <- sapply(c(0, 0.5, 1), function(x) length(intersect(which(all.geno[f.idx] == x), which(all.pheno[f.idx] < 34))))
        fmat <- rbind(num.high, num.low)
        #pheatmap(fmat, cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE)
        fp <- fisher.test(fmat)$p.value

        num.high <- sapply(c(0, 0.5, 1), function(x) length(intersect(which(all.geno[m.idx] == x), which(all.pheno[m.idx] > 34))))
        num.low <- sapply(c(0, 0.5, 1), function(x) length(intersect(which(all.geno[m.idx] == x), which(all.pheno[m.idx] < 34))))
        mmat <- rbind(num.high, num.low)
        #pheatmap(mmat, cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE)
        mp <- fisher.test(mmat)$p.value

        

    }
    cat("\n\n")
}

#f.geno <- round(geno_obj[f.idx,2,]*100)/100
#f.zbp1 <- f.geno[,zbp1.marker]
#f.alt <- f.geno[,gsub(".value", "", names(large.effect), fixed = TRUE)]
#stripchart(data_obj$pheno[f.idx,"X7h"]~f.zbp1*f.alt, vertical = TRUE, method = "jitter",
#    pch = 16, col = "gray")
```

## Interaction plots  {.tabset .tabset-fade .tabset-pills}

We also looked more broadly to see if there were interactions across 
all females (not just Zbp1 hets). The plots below show the interaction 
plots for all females.

```{r int_overall, fig.height = 5, fig.width = 10, results = "asis"}

full.geno <- round(geno_obj[f.idx,2,]*100)/100
full.pheno <- data_obj$pheno[f.idx,"X7h",drop=FALSE]


#pdf("~/Desktop/zbp1.pdf", width = 10, height = 5)
par(mfrow = c(1,2))
for(ch in 1:length(u_chr)){
    effect.chr <- u_chr[ch]
    max.effect <- which.max(-log10(test.pval[large.effect[which(large.effect.chr == effect.chr)]]))
    effect.idx <- which(data_obj$geno_names$locus == names(max.effect))
    names(effect.idx) <- names(max.effect)
    max.chr[[ch]] <- effect.idx
    max.loc <- data_obj$marker_location[effect.idx]
    marker.name <- paste0("Chr", effect.chr, "; ", signif(max.loc/1e6, 3), "Mb")
    cat("###", marker.name, "\n")
  
    plot_points(full.pheno, zbp1.geno[f.idx], full.geno[,effect.idx], pheno_name = "Body Temperature",
        marker1_label = "Zbp1", marker2_label = marker.name)

    plot_lines(full.pheno, zbp1.geno[f.idx], full.geno[,effect.idx], pheno_name = "Body Temperature",
        marker1_label = "Zbp1", marker2_label = marker.name, error_bars = "se")

    abline(h = 34)
    cat("\n\n")
}
#dev.off()
```

## LD blocks with significant effects {.tabset .tabset-fade .tabset-pills}

The LD blocks for the chromosomes with significant markers on them
are shown below. Colors indicate correlations between 0 (blue) and 
1 (red). Correlations below 0 are shown in gray.

Rough LD blocks are outlined based on medoid clustering of the 
genotype correlation matrix. The positions of the significant
markers are shown in blue, and the maximum marker is shown as a yellow
asterisk.

The x axis marks Mbp positions of the block boundaries as well as
the maximum marker. The blocks are pretty big because there seems
to be a fair bit of structure in the correlation matrix that makes
it hard to cluster.

The plot for Chr 10 looks a little funny because there were two 
regions of significance. 


```{r ld, results = "asis", fig.width = 7, fig.height = 6}
#pdf("~/Desktop/ld_test.pdf", width = 6, height= 6)
sig.blocks <- vector(mode = "list", length = length(u_chr))
names(sig.blocks) <- u_chr
for(ch in 1:length(u_chr)){
    cat("### Chr", u_chr[ch], "\n")
    chr.idx <- which(data_obj$chromosome == u_chr[ch])
    has.var <- which(apply(test.geno[,chr.idx], 2, var) > 0)
    geno.cor <- cor(test.geno[,chr.idx[has.var]])
    
    layout(matrix(c(1,2), nrow = 1), widths = c(1,0.2))
    par(mar = c(4,4,4,0))
    imageWithText(geno.cor, show.text = FALSE, 
        use.pheatmap.colors = TRUE, row.name = NULL, col.names = NULL,
        main = "", global.color.scale = TRUE, 
        global.min = 0, global.max = 1)
    max.idx <- which(colnames(test.geno)[chr.idx[has.var]] == names(max.chr[[ch]]))    

    testk <- test.pam.k(geno.cor, 2:10, plot.results = FALSE)
    cl.mean <- sapply(testk[[1]], mean)
    best.k <- as.numeric(names(which.max(cl.mean)))

    cl.geno <- pam(geno.cor, k = max(best.k, 4)) #use at least 4 clusters
    cl.mem <- merge_overlapping_clusters(cl.geno$clustering)
    u_cluster  <- unique(cl.mem)
    #plot(cl.geno$clustering)
    for(g in 1:length(u_cluster)){
        cl.idx <- which(cl.mem == g)
        draw.rectangle(min.x = min(cl.idx), max.x = max(cl.idx), 
            min.y = nrow(geno.cor)-max(cl.idx), max.y = nrow(geno.cor) - min(cl.idx),
            lwd = 3)
    }

    #show the significant markers in the LD plot
    sig.effects <- which(-log10(test.pval[chr.idx]) > nlog.thresh)
    sig.idx <- match(names(sig.effects), colnames(test.geno)[chr.idx])
    points(sig.idx, length(chr.idx) - sig.idx, pch = 16, col = "#2b8cbe")
    points(max.idx, length(chr.idx) - max.idx, pch = "*", cex = 2, col = "yellow")

    plot.dim <- par("usr")
    cl.idx <- lapply(u_cluster, function(x) which(cl.mem == x))
    cl.min.idx <- sapply(cl.idx, min)-1 #shift by one to merge split points
    cl.min.idx[which(cl.min.idx == 0)] <- 1
    cl.max.idx <- sapply(cl.idx, max)
    mark.at <- unique(sort(c(cl.min.idx, cl.max.idx, max.idx)))
    mark.pos <- round(signif(data_obj$marker_location[chr.idx[mark.at]], 3)/1e6)
    axis(1, at = mark.at, labels = mark.pos, las = 2)
    mtext("Position (Mbp)", side = 1, line = 2.5)
    best.pos <- round(signif(data_obj$marker_location[chr.idx[max.idx]], 3))/1e6
    mtext(paste0("Chr", u_chr[ch], "; Max at ", best.pos, " Mbp"), side = 3, line = 2.5)

    par(mar = c(15,2,6,3))
    imageWithTextColorbar(geno.cor, use.pheatmap.colors = TRUE, axis.line = 0,
        cex = 0.8, bar.lwd = 3, global.color.scale = TRUE, global.min = 0, 
        global.max = 1)


    cat("\n\n")
}
#dev.off()
```

## More complex inheritance?

The chromosome 3 locus looks like an interesting candidate and contains
the gene Mov10, which is predicted to be a helicase. Is it possible
that a more complex interaction can genetically resolve the animals
with high and low temperatures?

Below we explore this possibility using LASSO from the R package
glmnet. We fit a model that predicts numeric temperature in the 
female Zbp1 heterozygotes using genotype. The LASSO imposes a 
sparsity penalty so that we get as few markers as possible 
explaining the bimodal distribution.

```{r lasso, warning = FALSE}
#try LASSO

lasso.idx <- f.het #just the Zbp1 het females
#lasso.idx <- f.idx #all females
#lasso.idx <- 1:nrow(geno_obj) #all animals

#make test objects for the training using the right animals
lasso.geno <- round(geno_obj[lasso.idx,2,]*100)/100
lasso.pheno <- data_obj$pheno[lasso.idx,"X7h"] #numeric

#make a training data set that has the phenotype (either binary or quantitative) and markers
test.df <- data.frame(cbind("pheno" = lasso.pheno, lasso.geno))

test_acc <- function(pred, act){
    tn <- length(intersect(which(pred < 34), which(act < 34)))
    tp <- length(intersect(which(pred > 34), which(act > 34)))
    fp <- length(intersect(which(pred > 34), which(act < 34)))
    fn <- length(intersect(which(pred < 34), which(act > 34)))
    acc <- (tp + tn) / (tp + tn + fp + fn)
    return(acc)
}

trials <- 10
models <- vector(mode = "list", length = trials)
model.accuracy <- rep(NA, trials)
for(p in 1:trials){
    #print(p)
    sample <- sample(c(TRUE, FALSE), length(lasso.pheno), replace=TRUE, prob=c(0.7,0.3))
    train <- test.df[sample, ]
    test <- as.matrix(test.df[!sample,-1])
    x <- model.matrix(pheno ~ ., test.df)[,-1]
    y <- test.df$pheno
    model <- cv.glmnet(x, y, alpha = 1) #for numeric outcome
    models[[p]] <- model
    temp.pred <- predict(model, test)
    #plot(temp.pred, lasso.pheno[!sample]); abline(v = 34)
    model.accuracy[p] <- test_acc(temp.pred, lasso.pheno[!sample])
}

```

The plot below shows the distribution of coefficients for markers 
selected across all models.


```{r lasso_results, width = 10, height = 5}
coef.thresh = 0.1

acc.models <- which(model.accuracy == 1)
model.coef <- t(sapply(models[acc.models], function(x) as.vector(coef(x, s = "lambda.min"))))
markers.selected <- apply(model.coef, 1, function(x) which(x != 0))
total.markers <- setdiff(Reduce("union", markers.selected), 1)

boxplot(model.coef[,total.markers], ylab = "Coefficient", main = "non-zero LASSO coefficients", 
    xlab = "Index")
abline(h = 0)
abline(h = c(coef.thresh*-1, coef.thresh), lty = 2, col = "gray")
```

We selected all markers with absolute mean coefficients greater than 
`r coef.thresh`. The following table shows the names and locations
of these markers. 

The higher the threshold we use, the fewer markers we get, but the
less well temperature is predicted. This suggests that the bimodal
distribution seen in the Zbp1 hets is inherited in a complex manner.

The largest effect loci can be seen by sorting the table by coefficient.

```{r markers}
selected <- which(abs(colMeans(model.coef)) > coef.thresh)
selected.markers <- colnames(test.df)[selected[2:length(selected)]]
selected.idx <- match(selected.markers, data_obj$geno_name[[3]])
selected.chr <- data_obj$chromosome[selected.idx]
selected.pos <- data_obj$marker_location[selected.idx]
selected.coef <- colMeans(model.coef[,selected])
marker.df <- data.frame("marker" = selected.markers, "Chr" = selected.chr, 
    "Pos" = selected.pos, "Coef" = signif(colMeans(model.coef[,selected[2:length(selected)]]), 2))
datatable(marker.df)
```

The following plot shows a decomposition of the genotype matrix
using just these markers. Each point is an animal, and the color
indicates its body temperature at 7 hrs, where the bimodal 
distribution occurs. Blue dots indicate body temperatures below
34 degrees and red dots indicate body temperatures above 34 
degrees.

The first and second PCs together predict body temperature very well.


```{r pred, fig.height = 7, fig.width = 7}
selected.geno <- round(lasso.geno[,selected.idx], 2)
#pheatmap(selected.geno, cluster_rows = FALSE, cluster_cols = FALSE)

temp.cols <- colors.from.values(test.df$pheno, split.at.vals = TRUE, split.points = 34,
    col.scale = c("blue", "red"), grad.dir = "ends", light.dark = "d")

test.decomp <- plot.decomp(selected.geno, 
    cols = temp.cols, pc = min(c(3, length(selected.markers))))
```

```{r geno_img, eval = FALSE}
row.order <- order(test.df[,"pheno"])
temp.factor <- data.frame("high_temp" = as.factor(as.numeric(test.df[,"pheno"] > 34)))
rownames(temp.factor) <- rownames(test.df)

plot.mat <- selected.geno[row.order,]
colnames(plot.mat) <- apply(marker.df, 1, function(x) paste(x[2:3], collapse = "_"))
col.order <- order(marker.df[,"Coef"])
pheatmap(t(plot.mat), cluster_rows = FALSE, cluster_cols = FALSE, 
    annotation_col = temp.factor, show_colnames = FALSE, main = "ordered by genomic position")

pheatmap(t(plot.mat[,col.order]), cluster_rows = FALSE, cluster_cols = FALSE, 
    annotation_col = temp.factor, show_colnames = FALSE, main = "Ordered by effect size")


```

The following bar plot shows the correlation of each PC above
with body temperature. The second PC is very highly correlated
with body temperature.

```{r pc_cor, fig.width = 6, fig.height = 6}
barplot_with_num(apply(test.decomp$u, 2, function(x) signif(cor(x, lasso.pheno), 2)),
    main = "PC correlation with temperature at 7 hrs", ylab = "Pearson Correlation",
    names = paste0("PC", 1:ncol(test.decomp$u)), ylim = c(-0.6, 1))
```

The following plots show the PCs plotted against body temperature.
This suggests that this bimodal pattern is probably highly polygenic.
It takes `r length(selected.markers)` additional markers, to predict the 
bimodal split observed at the Zbp1 locus.

```{r temp_pc, fig.width = 12, fig.height = 4}
par(mfrow = c(1,3))
for(i in 1:ncol(test.decomp$u)){
    plot(test.decomp$u[,i], lasso.pheno, xlab = paste("PC", i), ylab = "Temperature", 
        pch = 16, main = paste("PC", i), col = temp.cols)
    abline(h = 34, v = 0)
}

```

What are the effects of positive and negative markers?
The plots below show the effects of the most positive 
and most negative markers.

They both have bimodal patterns, but the alternate
allele pushes the phenotype in opposite directions.

```{r single_marker, fig.width = 8, fig.height = 4}
par(mfrow = c(1,2))

max.pos <- which.max(marker.df[,"Coef"])
marker.name <- marker.df[max.pos,"marker"]
common.ind <- intersect(rownames(geno_obj), names(lasso.pheno))
boxplot(lasso.pheno[common.ind]~round(geno_obj[common.ind,2,marker.name], 2),
    xlab = "Genotype", ylab = "Temperature", main = "Most Positive Marker")
stripchart(lasso.pheno[common.ind]~round(geno_obj[common.ind,2,marker.name], 2),
    method = "jitter", add = TRUE, vertical = TRUE, pch = 16)


max.neg <- which.min(marker.df[,"Coef"])
marker.name <- marker.df[max.neg,"marker"]
common.ind <- intersect(rownames(geno_obj), names(lasso.pheno))
boxplot(lasso.pheno[common.ind]~round(geno_obj[common.ind,2,marker.name], 2),
    xlab = "Genotype", ylab = "Temperature", main = "Most Negative Marker")
stripchart(lasso.pheno[common.ind]~round(geno_obj[common.ind,2,marker.name], 2),
    method = "jitter", add = TRUE, vertical = TRUE, pch = 16)

```

## Non-Mendelian Ratios

It might also be valuable to look for interactions between the Zbp1
locus and other loci based on Mendelian ratios. Are genotypic combinations
at the Zpb1 locus prohibited by particular genotypes at other loci?

```{r interacting_loci}

#specify ind.idx to test only specific individuals, such 
#as males or females
test_two_ratios <- function(gene.name, ind.idx = NULL){

    if(is.null(ind.idx)){ind.idx <- 1:nrow(geno_obj)}

    gene.marker <- nearest_marker(data_obj, gene.name, gene.info)
    gene.geno <- round(geno_obj[ind.idx,2,gene.marker]*100)/100

    combo.geno.p <- rep(NA, dim(geno_obj)[3])
    for(i in 1:dim(geno_obj)[3]){
        test.marker <- round(geno_obj[ind.idx,2,i]*100)/100
        comb.freq <- table(data.frame(cbind(gene.geno, test.marker)))
        #pheatmap(comb.freq, cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE)
        test <- fisher.test(comb.freq)
        combo.geno.p[i] <- test$p.value
    }
    return(combo.geno.p)
}


get_sig_markers <- function(marker.stats, thresh){
    sig.idx <- which(marker.stats > thresh)
    sig.chr <- data_obj$chromosome[sig.idx]
    sig.pos <- data_obj$marker_location[sig.idx]
    sig.markers <- dimnames(geno_obj)[[3]][sig.idx]
    sig.table <- cbind(sig.idx,sig.markers, sig.chr, sig.pos)
    return(sig.table)
}

plot_combo_geno <- function(gene.name, comp.marker){
    gene.marker <- nearest_marker(data_obj, gene.name, gene.info)
    gene.geno <- round(geno_obj[,2,gene.marker]*100)/100
    comp.idx <- which(dimnames(geno_obj)[[3]] == comp.marker)
    comp.geno <- round(geno_obj[,2,comp.idx]*100)/100
    comb.freq <- table(data.frame(cbind(gene.geno, comp.geno)))
    layout.mat <- matrix(c(2,0,1,3), ncol = 2, byrow = TRUE)
    layout(layout.mat, widths = c(1, 0.4), heights = c(0.4, 1))
    par(mar = c(6,4,0,0))
    imageWithText(comb.freq, use.pheatmap.colors = TRUE, col.text.rotation = 0,
        col.text.shift = 0.3, row.text.shift = 0.3, cex = 1)
    par(mar = c(0,4,2,0))
    barplot(colSums(comb.freq), names = NA, ylab = comp.marker)
    par(mar = c(5,0,0,2))
    barplot(rev(rowSums(comb.freq)), names = NA, horiz = TRUE, xlab = gene.name)
}


perm_combo <- function(gene.name, comp.marker, nperm = 100, ind.idx = NULL){
    gene.marker <- nearest_marker(data_obj, gene.name, gene.info)
    gene.geno <- round(geno_obj[,2,gene.marker]*100)/100
    comp.idx <- which(dimnames(geno_obj)[[3]] == comp.marker)
    comp.geno <- round(geno_obj[,2,comp.idx]*100)/100
    comb.freq <- table(data.frame(cbind(gene.geno, comp.geno)))
    
    perm.freq <- vector(mode = "list", length = nperm)
    for(p in 1:nperm){
        test1 <- sample(gene.geno)
        test2 <- sample(comp.geno)
        perm.freq[[p]] <- table(data.frame(cbind(test1, test2)))
    }

    perm.avg <- Reduce("+", perm.freq)/nperm
    
    layout.mat <- matrix(c(2,4,1,3), ncol = 2, byrow = TRUE)
    layout(layout.mat, widths = c(1, 0.4), heights = c(0.4, 1))
    par(mar = c(6,4,0,0))
    imageWithText(perm.avg, use.pheatmap.colors = TRUE, col.text.rotation = 0,
        col.text.shift = 0.3, row.text.shift = 0.3, cex = 1)
    par(mar = c(0,4,2,0))
    barplot(colSums(perm.avg), names = NA, ylab = comp.marker)
    par(mar = c(5,0,0,2))
    barplot(rev(rowSums(perm.avg)), names = NA, horiz = TRUE, xlab = gene.name)
    plot.text("Expectation if\nIndependent")
}

```

The plot below shows the -log10(p values) for deviation from
expected frequencies with the Zbp1 locus. The large peak is the
Zbp1 locus itself. But there are some other small peaks, including
one on Chr 17 that might be interesting. 

Not shown are the stats for males and females separately.
They did not have any significant loci outside of the 
test marker region.

```{r interacting_Zbp1, fig.width = 10, fig.height = 5}
zbp1.inter <- test_two_ratios("Zbp1")

plot_stats_on_chr(-log10(zbp1.inter), data_obj$chromosome, mark.y = 3, 
    y.label = "-log10(p value)")

#zbp1.inter.f <- test_two_ratios("Zbp1", f.idx)
#plot_stats_on_chr(-log10(zbp1.inter.f), data_obj$chromosome, mark.y = 3, 
#    y.label = "-log10(p value)")
#mtext("Females", side = 3, outer = TRUE, line = -2.5)

#m.idx <- which(data_obj$pheno[,"sex"] == 1)
#zbp1.inter.m <- test_two_ratios("Zbp1", m.idx)
#plot_stats_on_chr(-log10(zbp1.inter.m), data_obj$chromosome, mark.y = 3, y.label = "-log10(p value)")
#mtext("Males", side = 3, outer = TRUE, line = -2.5)

zbp1.sig <- get_sig_markers(marker.stats = -log10(zbp1.inter), thresh = 3)
```

The following plot shows the frequencies of pairwise
genotypes for the marker nearest Zbp1 as well as a significant
marker on Chromosome 3. The MOLF double homozygotes are more
common than we would expect if the markers were assorting
independently.

```{r Zbp1_chr3, fig.width = 5, fig.height = 5}
plot_combo_geno("Zbp1", "S6J032449077")
```

The following plot shows the frequencies expected
if the loci were indepentent.

```{r Zbp1_chr3_perm, fig.width = 5, fig.height = 5}
perm_combo("Zbp1", "S6J032449077", nperm = 100)
```

The following plot shows the frequencies of pairwise
genotypes for the marker nearest Zbp1 as well as a significant
marker on Chromosome 17. This is not the Fshr locus. It is around
27 Mb. Fshr is at 89 Mb. The MOLF double homozygotes are more
common than we would expect if the markers were assorting
independently.

```{r Zbp1_chr17, fig.width = 5, fig.height = 5}
plot_combo_geno("Zbp1", "gUNC27721077")
```

The following plot shows the frequencies expected
if the loci were indepentent.

```{r Zbp1_chr17_perm, fig.width = 5, fig.height = 5}
perm_combo("Zbp1", "gUNC27721077", nperm = 100)
```

We can play the same game with the Fshr locus. The plot 
below shows the -log10(p values) for the independence of
markers with the marker nearest Fshr. The big peak shows 
us where Fshr is. There are no other loci that look as if
they are non-indepentendly assorting with the Fshr locus.

Not shown are the stats for males and females seprately.
There were no significant loci outside of the test marker
locus.

```{r interacting_Fshr, fig.width = 10, fig.height = 5}
fshr.inter <- test_two_ratios("Fshr")

plot_stats_on_chr(-log10(fshr.inter), data_obj$chromosome, mark.y = 3, 
    y.label = "-log10(p value)")

fshr.sig <- get_sig_markers(marker.stats = -log10(fshr.inter), thresh = 3)

#fshr.inter.f <- test_two_ratios("Fshr", f.idx)
#plot_stats_on_chr(-log10(fshr.inter.f), data_obj$chromosome, mark.y = 3, y.label = "-log10(p value)")

#fshr.inter.m <- test_two_ratios("Fshr", m.idx)
#plot_stats_on_chr(-log10(fshr.inter.m), data_obj$chromosome, mark.y = 3, y.label = "-log10(p value)")

```

### Full-genome scan

Let's do a full genome scan for non-Mendelian ratios.
For each marker on a chromosome we will test all markers
on other chromosomes.

```{r full_genome_non, eval = FALSE}

test_chr_non_mendelian <- function(chr, ind.idx = NULL){
   
   if(is.null(ind.idx)){ind.idx <- 1:nrow(geno_obj)}

    chr.idx <- which(data_obj$chromosome == chr)
    not.chr.idx <- which(data_obj$chromosome != chr)
    round.geno <- round(geno_obj[ind.idx,2,]*100)/100

    combo.geno.p <- matrix(NA, nrow = length(chr.idx), ncol = length(not.chr.idx))
    rownames(combo.geno.p) <- dimnames(geno_obj)[[3]][chr.idx]
    colnames(combo.geno.p) <- dimnames(geno_obj)[[3]][not.chr.idx]
    for(i in 1:length(chr.idx)){
        report.progress(i, length(chr.idx))
        for(j in 1:length(not.chr.idx)){
            test1 <- round.geno[,chr.idx[i]]
            test2 <- round.geno[,not.chr.idx[j]]
            comb.freq <- table(data.frame(cbind(test1, test2)))
            #pheatmap(comb.freq, cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE)
            test <- fisher.test(comb.freq)
            combo.geno.p[i, j] <- test$p.value
        }
        #plot_stats_on_chr(-log10(combo.geno.p[i,]), data_obj$chromosome, mark.y = 3, y.label = "-log10(p value)")
    }
    return(combo.geno.p)
}

full.genome.combo <- vector(mode = "list", length = 19)
names(full.genome.combo) <- 1:19
for(ch in 1:19){
    genome.combo.file <- here("Results", "non-mendelian", "genome", paste0("Chr", ch, ".RDS"))
    if(!file.exists(genome.combo.file)){
        test <- test_chr_non_mendelian(ch)
        saveRDS(test, genome.combo.file)
        cat("\n")
    }else{
        test <- readRDS(genome.combo.file)
    }
    full.genome.combo[[ch]] <- test
}


pheatmap(-log10(test), cluster_rows = FALSE, cluster_cols = FALSE, show_rownames = FALSE,
        show_colnames = FALSE)

```
